File: apps/api/app/schemas/stock_movement.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class StockMovementCreateRequest(BaseModel):
    catalog_item_id: str
    movement_type: str = Field(..., pattern="^(in|out)$")
    quantity: Decimal = Field(..., gt=0)
    movement_date: date
    notes: str | None = None
    reference_number: str | None = Field(default=None, max_length=128)
    procurement_id: str | None = None


class StockMovementRead(BaseModel):
    id: str
    catalog_item_id: str
    catalog_item_code: str | None = None
    catalog_item_name: str
    catalog_item_unit: str | None = None
    movement_type: str
    quantity: Decimal
    movement_date: date
    notes: str | None = None
    reference_number: str | None = None
    procurement_id: str | None = None
    balance_after: Decimal
    created_by_user_id: str
    created_by_email: str | None = None
    created_at: datetime
    updated_at: datetime


class StockMovementListResponse(BaseModel):
    items: list[StockMovementRead]
    total: int
    limit: int
    offset: int


class StockBalanceRead(BaseModel):
    catalog_item_id: str
    catalog_item_code: str | None = None
    catalog_item_name: str
    catalog_item_unit: str | None = None
    quantity: Decimal
    min_stock_level: Decimal | None = None
    is_below_min: bool = False


class LowStockListResponse(BaseModel):
    items: list[StockBalanceRead]
    total: int

File: apps/web/types/stock-movement.ts
export type StockMovementType = 'in' | 'out'

export interface StockMovementItem {
  id: string
  catalog_item_id: string
  catalog_item_code?: string | null
  catalog_item_name: string
  catalog_item_unit?: string | null
  movement_type: StockMovementType
  quantity: string | number
  movement_date: string
  notes?: string | null
  reference_number?: string | null
  procurement_id?: string | null
  balance_after: string | number
  created_by_user_id: string
  created_by_email?: string | null
  created_at: string
  updated_at: string
}

export interface StockMovementInput {
  catalog_item_id: string
  movement_type: StockMovementType
  quantity: string
  movement_date: string
  notes?: string
  reference_number?: string
  procurement_id?: string
}

export interface StockMovementListFilters {
  movement_type?: StockMovementType | ''
  catalog_item_id?: string
  q?: string
  reference_number?: string
  movement_date_from?: string
  movement_date_to?: string
  limit?: number
  offset?: number
}

export interface StockMovementListResponse {
  items: StockMovementItem[]
  total: number
  limit: number
  offset: number
}

export interface StockBalanceItem {
  catalog_item_id: string
  catalog_item_code?: string | null
  catalog_item_name: string
  catalog_item_unit?: string | null
  quantity: string | number
  min_stock_level?: string | number | null
  is_below_min: boolean
}

export interface LowStockListResponse {
  items: StockBalanceItem[]
  total: number
}
